src/systems/optical_tactile_plugin/Visualization.cc

```diff
@@ -0,0 +1,239 @@
+  this->forceMarkerMsg.mutable_material()->mutable_diffuse()->set_g(1);
+  this->forceMarkerMsg.mutable_material()->mutable_diffuse()->set_b(0);
+  this->forceMarkerMsg.mutable_material()->mutable_diffuse()->set_a(1);
+  this->forceMarkerMsg.mutable_lifetime()->set_sec(this->cameraUpdateRate);
+
+  this->sensorMarkerMsg.mutable_material()->mutable_ambient()->set_r(0.5);
+  this->sensorMarkerMsg.mutable_material()->mutable_ambient()->set_g(0.5);
+  this->sensorMarkerMsg.mutable_material()->mutable_ambient()->set_b(0.5);
+  this->sensorMarkerMsg.mutable_material()->mutable_ambient()->set_a(0.75);
+  this->sensorMarkerMsg.mutable_material()->mutable_diffuse()->set_r(0.5);
+  this->sensorMarkerMsg.mutable_material()->mutable_diffuse()->set_g(0.5);
+  this->sensorMarkerMsg.mutable_material()->mutable_diffuse()->set_b(0.5);
+  this->sensorMarkerMsg.mutable_material()->mutable_diffuse()->set_a(0.75);
+  this->sensorMarkerMsg.mutable_lifetime()->set_sec(0);
+  this->sensorMarkerMsg.mutable_lifetime()->set_nsec(0);
+
+  // Set scales
+  ignition::msgs::Set(this->positionMarkerMsg.mutable_scale(),
+    ignition::math::Vector3d(this->contactRadius, this->contactRadius,
+    this->contactRadius));
+
+  ignition::msgs::Set(this->forceMarkerMsg.mutable_scale(),
+    ignition::math::Vector3d(this->forceRadius, this->forceRadius,
+    this->forceLength));
+
+  ignition::msgs::Set(this->sensorMarkerMsg.mutable_scale(), _sensorSize);
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginVisualization::VisualizeSensor(
+  ignition::math::Pose3f &_sensorPose)
+{
+  this->sensorMarkerMsg.set_action(ignition::msgs::Marker::ADD_MODIFY);
+  this->sensorMarkerMsg.set_id(1);
+
+  this->sensorMarkerMsg.mutable_pose()->mutable_position()->set_x(
+    _sensorPose.Pos().X());
+  this->sensorMarkerMsg.mutable_pose()->mutable_position()->set_y(
+    _sensorPose.Pos().Y());
+  this->sensorMarkerMsg.mutable_pose()->mutable_position()->set_z(
+    _sensorPose.Pos().Z());
+  this->sensorMarkerMsg.mutable_pose()->mutable_orientation()->set_x(
+    _sensorPose.Rot().X());
+  this->sensorMarkerMsg.mutable_pose()->mutable_orientation()->set_y(
+    _sensorPose.Rot().Y());
+  this->sensorMarkerMsg.mutable_pose()->mutable_orientation()->set_z(
+    _sensorPose.Rot().Z());
+  this->sensorMarkerMsg.mutable_pose()->mutable_orientation()->set_w(
+    _sensorPose.Rot().W());
+
+  this->node.Request("/marker", this->sensorMarkerMsg);
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginVisualization::VisualizeNormalForce(
+  uint64_t &_i, uint64_t &_j,
+  ignition::math::Vector3f &_position,
+  ignition::math::Vector3f &_normalForce,
+  ignition::math::Pose3f &_sensorWorldPose)
+{
+  // Visualization is downsampled according to plugin parameter
+  // <visualization_resolution> if <visualize_forces> is set to true
+
+  // Check if _normalForce has to be visualized
+  bool visualizeForce = (_i % this->visualizationResolution == 0 &&
```

Comment:
> Instead of doing modulo at all, you can simply increment the i and j loops by visualizationResolution. That should eliminate all of the modulo operations and would not need any conditional checks.

I like this one, but regarding normal forces we have computation (which will be published in a message in the future) and visualization. Does it make sense to not computing the normal forces for the whole range of the depth camera? And if it does, should that resolution be the same of the visualization?

Reply:
Do you mean the normals represented as an RGB image? Yes, I think we will need to sacrifice the resolution at the benefit of performance. It should also be more intuitive to the user, that the resolution they specify applies to all the resolutions.

For the other messages (ones that are not an RGB image), the resolution of the normals is independent of the original depth resolution. The normal computation is only using the raw depth values, right? At that, it is only using the immediate neighboring pixels, with indices changing at a magnitude of 1 (i.e. i +/- 1, j +/- 1). So it shouldn't matter what resolution the result is, the computation should be the same.

Reply:
> So it shouldn't matter what resolution the normals are computed on, the pixel operations on the raw depth values will be the same.

Correct, just asked in case we wouldn't want different visualization & computation resolutions.

Reply:
e216bb2